### PR TITLE
Reactive value generation

### LINKED ISSUE #1099: Reactive ValueGenerator<> for use with @GeneratorType?
Hello! 
I have a column I need generated from a SQL snippet during session. It uses SELECT to grab a db specific global variable and stuff it into a column on insert using     
```java 
@GeneratorType( type = GrabGlobalDbVarGenerator.class, when = GenerationTime.INSERT)
    @Column(updatable = false, nullable = false)
    private long dbFetchedGlobal;
```
How can I do this in Hibernate Reactive?  `ValueGenerator<Long>`  doesn't use reactive Session.

I can `session.unwrap(ReactiveSession.class)` but there's no return of `CompletionStage<Long>` so I don't think it will work properly

Can you point me in right direction if this is supported?

### PR DESCRIPTION
**Pull request by DavideD**
Fixes #1099

### CODE REVIEW

**Comment by gavinking** [comment]
OK, this patch is bigger than expected, so can you give me until tomorrow evening to look it over?

**Comment by DavideD** [comment]
Sure.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java [hunk 1]
```diff
@@ -201,102 +211,119 @@ default ReactiveConnection getReactiveConnection(SharedSessionContractImplemento
...
-					).thenApply( v -> id )
-			);
+	default CompletionStage<Void> reactivePreInsertInMemoryValueGeneration(Object[] fields, Object object, SharedSessionContractImplementor session) {
+		CompletionStage<Void> stage = voidFuture();
+		if ( getEntityMetamodel().hasPreInsertGeneratedValues() ) {
+			final InMemoryValueGenerationStrategy[] strategies = getEntityMetamodel().getInMemoryValueGenerationStrategies();
+			for ( int i = 0; i < strategies.length; i++ ) {
+				final int index = i;
+				final InMemoryValueGenerationStrategy strategy = strategies[i];
+				if ( strategy != null && strategy.getGenerationTiming().includesInsert() ) {
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java:275]: What happens to in-memory (nonreactive) value generators?

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/tuple/MutinyValueGenerator.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * </p>
+ *
+ * @see StageValueGenerator
+ */
+public interface MutinyValueGenerator<T> extends ValueGenerator<T> {
+
+	@Override
+	default T generateValue(Session session, Object owner) {
+		throw new UnsupportedOperationException( "Use generateValue(Mutiny.Session, Object) instead" );
+	}
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/tuple/MutinyValueGenerator.java:28]: Ugh. This isn't very nice.

**Comment by gavinking** [comment]
So I think we need to find some way of removing the inheritance from `ValueGeneration`.

That's a problem I had not noticed.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java (hunk 1 above)
Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java:275]: Not sure I understand the question. But I've just moved the existing code

**Comment by DavideD** [comment]
Yeah, I agree, but I cannot think of a better solution right now.

**Comment by gavinking** [comment]
Right, neither can I.

**Comment by DavideD** [comment]
Maybe having abstract classes instead of interfaces is slightly better in this scenario: 2998ff0f5f3ff82d1d787fbb836a5ef8b27c4442

**Comment by gavinking** [comment]
Because it lets you redeclare the method `final`, you mean?

**Comment by DavideD** [comment]
Yes

**Comment by DavideD** [comment]
I'm also going to add the `@Incubating` annotation to clarify that's a temporary solution.
This way people people have at least a workaround for this feature.

**Comment by gavinking** [comment]
> I'm also going to add the `@Incubating` annotation to clarify that's a temporary solution.
> This way people have at least a workaround for this feature.

That's a good idea.

**Comment by DavideD** [comment]
Done. I'm going to merge it as it is if there is no strong opposition

**Comment by gavinking** [comment]
My only question was: does it break in-memory impls of `ValueGeneration`, or do they still work?

**Comment by DavideD** [comment]
How can I test it? All our tests are passing ...

**Comment by gavinking** [comment]
Well, the `@CurrentUser` annotation is our test for that.

**Comment by gavinking** [comment]
Sorry, I mean `@GeneratorType(type = LoggedUserGenerator.class, when = GenerationTime.ALWAYS)`

**Comment by DavideD** [comment]
I've moved the code around but I don't think I've changed anything of the original logic (not on purpose at least).
The core of the change is the method [ReactiveAbstractEntityPersister#generateValue](https://github.com/DavideD/hibernate-reactive/blob/b235ae9e3071c6bbd41a6046cab84fc53d1da622/hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java#L753).

If a user will use the new reactive classes, Hibernate Reactive will call the reactive methods. Implementations of `ValueGenerator` [will still work](https://github.com/DavideD/hibernate-reactive/blob/b235ae9e3071c6bbd41a6046cab84fc53d1da622/hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java#L772).

We test the new classes via `GeneratedProperty*Test` but we also have a test for `ValueGenerator` via `TimestampTest` because it tests the use of the annotations `@CreationTimestamp` and `@UpdateTimestamp` (they use a valuegenerator underneath). Maybe adding a direct test for ValueGenerator would be wise though.

The drawback of allowing an implementation of `ValueGenerator` is that a user can access the ReactiveSession, something that normally would be hidden.

Overall I think our tests cover most cases (I was pleasantly surprised when I've seen the test suite fails when I've started to work on this) but  if you have some other test in mind, let me know.

**Comment by DavideD** [comment]
Merged.
This should be good enough for now.

**Pull request merged**